src/gre.rs

```diff
@@ -0,0 +1,289 @@
+use crate::eth::EtherType;
+
+/// Represents a Generic Routing Encapsulation (GRE) header as defined in RFC 2784.
+///
+/// GRE is a tunneling protocol that encapsulates a wide variety of network layer
+/// protocols inside virtual point-to-point links over an Internet Protocol network.
+///
+/// This struct represents the maximum possible size of the GRE header, including
+/// the optional checksum and reserved fields. The presence of these optional fields
+/// is determined by the `checksum_present` flag. The `header_len()` method can be
+/// used to determine the actual length of the header at runtime (4 or 8 bytes).
+///
+/// For more details, see RFC 2784: https://www.rfc-editor.org/rfc/rfc2784.html
+/// 
+/// /// A struct containing the optional checksum and reserved fields.
+
+#[repr(C, packed)]
+#[derive(Debug, Copy, Clone)]
+pub struct GreOptionalFields {
+    /// This field is only valid if the Checksum Present flag is set.
+    pub check: [u8; 2],
+    /// A reserved field for future use, which MUST be transmitted as zero (optional).
+    /// This field is only present if the Checksum Present flag is set.
+    pub _reserved1: [u8; 2],
+}
+
+/// A union representing the 4 bytes that can either be optional GRE fields
+/// or the start of the payload data.
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub union GreDataUnion {
+    /// The interpretation when checksum_present() is true.
+    fields: GreOptionalFields,
+    /// The interpretation when checksum_present() is false.
+    payload_start: [u8; 4],
```

Comment:
I could be wrong here, but setting this to 0 makes more sense since we don't want to represent the packet payload data within this union.

```suggestion
    payload_start: [u8; 0],
```

Reply:
The union still has the length of 4 either way though, doesn't it? We're ignoring it in the main header struct and reading the length as 4 in the case of no checksum, but I think regardless it'll still include the first 4 of the payload.